# Face Recognition Project With OpenCV

Since we know that the Haar Cascade Classifier is nothing but Viola-Jones Algorithm. Now let's download the Weiler Jones Algorithm that is Haar Cascade Classifier for face detection. Go to https://github.com/opencv/opencv/tree/master/data/haarcascades. Here you can see the bunch of Cascade Classifiers available which will actually do the specific task. There are many pre-defined Cascade Classifier that are available. What you need to do is just download it and write some code doing the object detection. For our project download haarcascade_frontalface_default.xml.

## Load Image And Cascade Classifier

In [1]:
import numpy as np
import cv2

In [2]:
img = cv2.imread('faces.jpg')


cv2.imshow('face',img)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [3]:
face_cascade = cv2.CascadeClassifier('./model/haarcascade_frontalface_default.xml')

## Apply Viola-Jones Framework(Cascade Classsifier) To Image

Remember we can't directly directly apply the Cascade Classifier to our image. We need to convert the image into a gray scale
and then we need to apply to the Cascade Classifier to get the bounding boxes.

In [4]:
# step-1: Convert image into gray scale
gray = cv2.cvtColor(img,cv2.COLOR_BGR2GRAY)

# step-2: apply gray scale image to cascasde classifier
box,detections = face_cascade.detectMultiScale2(gray)
# detections/number of detections basically refers to the minimum number of neighbours, means number of neighbours
# that was detected.

In [5]:
box

array([[321,  84, 385, 385],
       [697, 107, 248, 248],
       [ 88,  75, 316, 316],
       [894, 573,  52,  52]])

The four rows indicate that we have the four faces

In [6]:
detections

array([35, 37, 10,  4])

This says row1 of box has 35 neighbours detected and so on...
Let's set some threshold value for minimum neighbours.

In [7]:
# step-1: Convert image into gray scale
gray = cv2.cvtColor(img,cv2.COLOR_BGR2GRAY)

# step-2: apply gray scale image to cascasde classifier
box,detections = face_cascade.detectMultiScale2(gray, minNeighbors=8)
# detections/number of detections basically refers to the minimum number of neighbours, means number of neighbours
# that was detected.

In [8]:
box

array([[321,  84, 385, 385],
       [ 88,  75, 316, 316],
       [697, 107, 248, 248]])

In [9]:
detections

array([35, 10, 37])

The meaning of these boxes is that 321 and 81, these two values represents the xy-position and 385 represents width of the 
box and the next 385 represents the height of the box or precisely the width of the face and height of the face.

## Draw Bounding Boxes

In [10]:
image = img.copy()
# step-1: Convert image into gray scale
gray = cv2.cvtColor(img,cv2.COLOR_BGR2GRAY)
# step-2: apply gray scale image to cascasde classifier
box,detections = face_cascade.detectMultiScale2(gray,minNeighbors=8)
# step-3: drawing bounding box
for x,y,w,h in box:

    cv2.rectangle(image,(x,y),(x+w,y+h),(0,255,0),1)

cv2.imshow('face detection',image)
cv2.waitKey(0)
cv2.destroyAllWindows()

## Face Detection Function

In [11]:
def face_detection(img):

    image = img.copy()
    # step-1: Convert image into gray scale
    gray = cv2.cvtColor(img,cv2.COLOR_BGR2GRAY)
    # step-2: apply gray scale image to cascasde classifier
    box,detections = face_cascade.detectMultiScale2(gray,minNeighbors=8)
    # step-3: drawing bounding box
    for x,y,w,h in box:

        cv2.rectangle(image,(x,y),(x+w,y+h),(0,255,0),1)
        
    return image

In [12]:
img_detect = face_detection(img)


cv2.imshow('face detection',img_detect)
cv2.waitKey(0)
cv2.destroyAllWindows()

## Real Time Face Detection

In [13]:
cap = cv2.VideoCapture(0)

face_cascade = cv2.CascadeClassifier('./model/haarcascade_frontalface_default.xml')

while True:
    ret, frame = cap.read()
    if ret == False:
        break
        
    img_detect = face_detection(frame)
    
    cv2.imshow('Real Time Face Detection',img_detect)
    if cv2.waitKey(1) == ord('q'):
        break
        
        
cap.release()
cv2.destroyAllWindows()